src/CoreWCF.Http/src/CoreWCF/Channels/HttpChannelHelpers.cs

```diff
@@ -121,74 +123,81 @@ private void ThrowMaxReceivedMessageSizeExceeded()
-                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(
-                        new ProtocolException(SR.MessageXmlProtocolError, xmlException));
-                }
+                return await _messageEncoder.ReadMessageAsync(buffer, _bufferManager, ContentType);
             }
-            finally
+            catch (XmlException xmlException)
             {
-                inputStream.Close();
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(
+                    new ProtocolException(SR.MessageXmlProtocolError, xmlException));
             }
         }
 
         private async Task<Message> ReadBufferedMessageAsync(Stream inputStream)
         {
-            ArraySegment<byte> messageBuffer = GetMessageBuffer();
-            byte[] buffer = messageBuffer.Array;
-            int offset = 0;
-            int count = messageBuffer.Count;
-
-            while (count > 0)
+            var reader = PipeReader.Create(inputStream);
+            try
             {
-                int bytesRead = await inputStream.ReadAsync(buffer, offset, count);
-                if (bytesRead == 0) // EOF
+                ReadOnlySequence<byte> buffer;
+                while (true)
                 {
-                    if (ContentLength != -1)
+                    ReadResult result = await reader.ReadAsync();
+                    buffer = result.Buffer;
+
+                    reader.AdvanceTo(buffer.Start, buffer.End);
+
+                    if (result.IsCompleted)
                     {
-                        throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(
-                            new ProtocolException(SR.HttpContentLengthIncorrect));
+                        break;
                     }
-
-                    break;
                 }
-                count -= bytesRead;
-                offset += bytesRead;
-            }
 
-            return await DecodeBufferedMessageAsync(new ArraySegment<byte>(buffer, 0, offset), inputStream);
+                return await DecodeBufferedMessageAsync(buffer);
+            }
+            finally
+            {
+                await reader.CompleteAsync();
+            }
         }
 
         private async Task<Message> ReadChunkedBufferedMessageAsync(Stream inputStream)
         {
+            var reader = PipeReader.Create(inputStream, new StreamPipeReaderOptions(bufferSize: ConnectionOrientedTransportDefaults.ConnectionBufferSize));
             try
             {
-                return _messageEncoder.ReadMessage(await BufferMessageStreamAsync(inputStream, _bufferManager, _settings.MaxBufferSize), _bufferManager, ContentType);
+                ReadOnlySequence<byte> buffer;
+                while (true)
+                {
+                    ReadResult result = await reader.ReadAsync();
+                    buffer = result.Buffer;
+
+                    if (ContentLength == -1 && buffer.Length >= _settings.MaxBufferSize)
+                    {
```

Comment:
This method is only called when ContentLength == -1, no need to double check. I see you are doing the correct length check here. It looks like we don't need separate methods for ReadBufferedMessageAsync and ReadChunkedBufferedMessageAsync as the implementation is looking like it should be the same now, unless I missed something.

Reply:
I think this is now fixed.